### PR TITLE
samples: Spark connector word count sample

### PR DESCRIPTION
**Pull request by jiangmichaellll**
This adds the word count sample and integration test.

1. We will need to change the values inside integration test with java-doc-sample visible cluster and topic etc.
2. Don't make samples as a submodule, because that would require top level pom to be packaged as `pom` instead of `jar`, as a result, integration test won't be triggered directly from top level `mvn test`, instead it needs to go into `samples` folder and then `mvn test`. I saw java-pubsublite has this structure so I assume it will work in kokoro.
3. integration test needs to have a valid maven home to work, I assumed that kokoro should have access to that, if not, it will need another lib

### CODE REVIEW

File: samples/snapshot/src/test/java/pubsublite/spark/SampleIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  @Test
+  public void test() throws Exception {
+    // Maven package into jars
+    mavenPackage(WORKING_DIR);
+    mavenPackage(WORKING_DIR + "/samples");
+
+    // Upload to GCS
+    Storage storage =
+        StorageOptions.newBuilder().setProjectId(PROEJCT_ID.value()).build().getService();
+    BlobId blobId = BlobId.of(BUCKET_NAME, SAMPLE_JAR_NAME);
```
Comment (anguillanneuf) [inline samples/snapshot/src/test/java/pubsublite/spark/SampleIntegrationTest.java:115]: Avoid parallel uploads from parallel tests.

```suggestion
    BlobId blobId = BlobId.of(BUCKET_NAME, suffix, SAMPLE_JAR_NAME);
```

File: samples/snapshot/src/test/java/pubsublite/spark/SampleIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+    storage.create(blobInfo, Files.readAllBytes(Paths.get(SAMPLE_JAR_LOC)));
+    blobId = BlobId.of(BUCKET_NAME, CONNECTOR_JAR_NAME);
```
Comment (anguillanneuf) [inline samples/snapshot/src/test/java/pubsublite/spark/SampleIntegrationTest.java:118]:
```suggestion
    blobId = BlobId.of(BUCKET_NAME, suffix, CONNECTOR_JAR_NAME);
```

Ditto.

**Review by anguillanneuf — COMMENTED** [review]
Let's chat offline.

File: samples/snapshot/src/test/java/pubsublite/spark/SampleIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.spark_project.guava.collect.ImmutableList;
+
+public class SampleIntegrationTest {
+
+  private static final CloudRegion CLOUD_REGION = CloudRegion.of("us-central1");
+  CloudZone CLOUD_ZONE = CloudZone.of(CLOUD_REGION, 'b');
+  ProjectId PROEJCT_ID = ProjectId.of("pubsub-lite-load-tests");
+  TopicName TOPIC_NAME = TopicName.of("jiangmichael-sample-integration");
```
[the remark this answers is not in the document]
Comment (jiangmichaellll) [reply samples/snapshot/src/test/java/pubsublite/spark/SampleIntegrationTest.java]: topic is fixed.

File: samples/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+   ```
+7. Set Dataproc region
+   ```
+   gcloud config set dataproc/region $REGION
+   ```
+<!-- TODO: set up bots to update jar version -->
+7. Run the sample in Dataproc
+   ```
+   gcloud dataproc jobs submit spark --cluster=$CLUSTER_NAME \
+      --jars=$BUCKET/pubsublite-spark-snapshot-1.0.21.jar,$BUCKET/pubsublite-spark-sql-streaming-with-dependencies-0.1.0-SNAPSHOT \
```
Comment (anguillanneuf) [inline samples/README.md:65]:
```suggestion
      --jars=$BUCKET/pubsublite-spark-snapshot-1.0.21.jar,$BUCKET/pubsublite-spark-sql-streaming-with-dependencies-0.1.0-SNAPSHOT.jar \
```

File: samples/snippets/src/main/java/pubsublite/spark/AdminUtils.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            .setName(topicPath.toString())
+            .build();
+
+    AdminClientSettings adminClientSettings =
+        AdminClientSettings.newBuilder().setRegion(CloudRegion.of(cloudRegion)).build();
+
+    try (AdminClient adminClient = AdminClient.create(adminClientSettings)) {
+      Topic response = adminClient.createTopic(topic).get();
+      System.out.println(response.getAllFields() + "created successfully.");
+    }
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsublite/spark/AdminUtils.java:87]: Can you catch and ignore `com.google.api.gax.rpc.AlreadyExistsException` errors here.

File: samples/snippets/src/main/java/pubsublite/spark/AdminUtils.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            .setTopic(topicPath.toString())
+            .build();
+
+    AdminClientSettings adminClientSettings =
+        AdminClientSettings.newBuilder().setRegion(CloudRegion.of(cloudRegion)).build();
+
+    try (AdminClient adminClient = AdminClient.create(adminClientSettings)) {
+      Subscription response = adminClient.createSubscription(subscription).get();
+      System.out.println(response.getAllFields() + "created successfully.");
+    }
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsublite/spark/AdminUtils.java:130]: Ditto. Can you catch and ignore `com.google.api.gax.rpc.AlreadyExistsException` errors here.

File: samples/README.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+```
+PROJECT_ID=your-project-id
+REGION=us-central1 # or your region
+ZONE_ID=b # or your zone id
+TOPIC_ID=test-topic # or your topic id to create
+SUBSCRIPTION_ID=test-subscrciption # or your subscription to create
+PARTITIONS=1 # or your number of partitions to create
+CLUSTER_NAME=waprin-spark7 # or your Dataproc cluster name to create
+BUCKET=gs://your-gcs-bucket
+SUBSCRIPTION_PATH=projects/$PROJECT_NUMBER/locations/$REGION-$ZONE_ID/subscriptions/$SUBSCRIPTION_ID
```
Comment (anguillanneuf) [inline samples/README.md:21]:
```suggestion
SUBSCRIPTION_PATH=projects/$PROJECT_ID/locations/$REGION-$ZONE_ID/subscriptions/$SUBSCRIPTION_ID
```

File: samples/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+   gcloud config set dataproc/region $REGION
+   ```
+<!-- TODO: set up bots to update jar version -->
+7. Run the sample in Dataproc
+   ```
+   gcloud dataproc jobs submit spark --cluster=$CLUSTER_NAME \
+      --jars=$BUCKET/pubsublite-spark-snapshot-1.0.21.jar,$BUCKET/pubsublite-spark-sql-streaming-with-dependencies-0.1.0-SNAPSHOT \
+      --class=pubsublite.spark.WordCount -- $SUBSCRIPTION_PATH
+   ```
+
```
Comment (anguillanneuf) [inline samples/README.md:68]: Please add clean up steps to delete Lite topic, subscription, Dataproc cluster, GCS bucket. 
```suggestion
## Cleaning up
1. Delete Pub/Sub Lite topic and subscription.
```
gcloud pubsub lite-subscriptions delete $SUBSCRIPTION_ID --zone=$REGION-$ZONE_ID
gcloud pubsub lite-topics delete $TOPIC_ID --zone=$REGION-$ZONE_ID
```
1. Delete GCS bucket.
```
gsutil -m rm -rf $BUCKET_NAME
```
1. Delete Dataproc cluster.
```
gcloud dataproc clusters delete $CLUSTER_NAME --region=$REGION
```

File: samples/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+3. Create a Dataproc cluster
+   ```
+   gcloud dataproc clusters create $CLUSTER_NAME --region=$REGION --zone="$REGION-$ZONE_ID" --image-version=1.5-debian10 --scopes=cloud-platform
+   ```
+4. Package sample jar
+   ```
+   mvn clean package -Dmaven.test.skip=true
+   ```
+<!-- TODO: set up bots to update jar version, also provide link to maven central --> 
+5. Download pubsublite-spark-sql-streaming-with-dependencies-0.1.0.jar from Maven Central and set PUBSUBLITE_SPARK_SQL_STREAMING_JAR_LOCATION environment variable.
```
Comment (anguillanneuf) [inline samples/README.md:49]:
```suggestion
5. Download `pubsublite-spark-sql-streaming-with-dependencies-0.1.0.jar` from Maven Central and set `PUBSUBLITE_SPARK_SQL_STREAMING_JAR_LOCATION` environment variable.
```

File: samples/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+<!-- TODO: set up bots to update jar version -->
+6. Create GCS bucket and upload both pubsublite-spark-sql-streaming-with-dependencies-0.1.0 jar and the sample jar onto GCS
```
Comment (anguillanneuf) [inline samples/README.md:51]:
```suggestion
6. Create GCS bucket and upload both `pubsublite-spark-sql-streaming-with-dependencies-0.1.0.jar` and the sample jar onto GCS.
```

File: samples/README.md (hunk 7 above)
Comment (jiangmichaellll) [reply samples/README.md:21]: I think this needs a project number, I added that env variable above.

Comment (anguillanneuf) [reply samples/README.md:21]: I gave it a try using PROJECT_ID and it seemed to have worked.

File: samples/README.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+   gsutil mb $BUCKET
+   gsutil cp snapshot/target/pubsublite-spark-snapshot-1.0.21.jar $BUCKET
+   gsutil cp $PUBSUBLITE_SPARK_SQL_STREAMING_JAR_LOCATION $BUCKET
+   ```
+7. Set Dataproc region
+   ```
+   gcloud config set dataproc/region $REGION
+   ```
+<!-- TODO: set up bots to update jar version -->
+7. Run the sample in Dataproc
```
Comment (anguillanneuf) [inline samples/README.md:62]:
```suggestion
8. Run the sample in Dataproc
```

File: samples/snapshot/pom.xml [hunk 12]
```diff
@@ -4,7 +4,7 @@
   <groupId>com.google.cloud</groupId>
   <artifactId>pubsublite-spark-snapshot</artifactId>
   <packaging>jar</packaging>
-  <name>Google Pub/Sub Lite Spark Connector Snapshot Samples</name>
+  <name>Google Pub/Sub Lite Spark Connector Snapshot</name>
```
Comment (anguillanneuf) [inline samples/snapshot/pom.xml:7]: What's the reason to remove "Samples" here? I'd leave it in.

File: samples/README.md (hunk 7 above)
Comment (jiangmichaellll) [reply samples/README.md:21]: hm I think we added this support recently. Anyway I switched back to project number in all readme user samples.

File: samples/README.md (hunk 11 above)
Comment (jiangmichaellll) [reply samples/README.md:62]: good catch.

File: samples/snapshot/pom.xml (hunk 12 above)
Comment (jiangmichaellll) [reply samples/snapshot/pom.xml:7]: I am keeping snippets the same... ok, let me keep it.

**Review by anguillanneuf — APPROVED** [review]
LGTM. Thanks for adding these samples!

**Pull request merged**